src/main/ApplicationImpl.cpp

```diff
@@ -549,27 +550,224 @@ ApplicationImpl::joinAllThreads()
+                           "for database (maximum {})"),
+                *userSpecifiedSeqNum, std::numeric_limits<int32_t>::max()));
+        }
+
+        if (*userSpecifiedSeqNum <= startLedgerSeq)
+        {
+            throw std::invalid_argument(fmt::format(
+                FMT_STRING("Invalid manual close ledger sequence number {} "
+                           "(must exceed current sequence number {})"),
+                *userSpecifiedSeqNum, startLedgerSeq));
+        }
+    }
+
+    return userSpecifiedSeqNum ? *userSpecifiedSeqNum : nextLedgerSeq;
+}
+
+void
+ApplicationImpl::setManualCloseVirtualTime(
+    optional<TimePoint> const& userSpecifiedCloseTime)
+{
+    TimePoint const maxSystemTimePointSeconds =
+        std::chrono::duration_cast<std::chrono::seconds>(
+            VirtualClock::system_time_point::duration::max())
+            .count();
+    TimePoint const maxTimeT = std::numeric_limits<std::time_t>::max();
+    TimePoint const maxCloseTime =
+        std::min(maxSystemTimePointSeconds, maxTimeT) -
+        Herder::MAX_TIME_SLIP_SECONDS.count();
+
+    TimePoint const lastCloseTime = getLedgerManager()
+                                        .getLastClosedLedgerHeader()
+                                        .header.scpValue.closeTime;
+    if (lastCloseTime >= maxCloseTime)
+    {
+        throw std::invalid_argument(
+            fmt::format(FMT_STRING("Close time already at max ({}); can not "
+                                   "close further ledgers"),
+                        maxCloseTime));
+    }
+
+    auto systemNow = getClock().system_now();
+    TimePoint secondsSinceEpoch =
+        std::chrono::duration_cast<std::chrono::seconds>(
+            systemNow.time_since_epoch())
+            .count();
+    if (secondsSinceEpoch > maxCloseTime)
+    {
+        throw std::invalid_argument(
+            fmt::format(FMT_STRING("System time already past max ({}); can not "
+                                   "close further ledgers"),
+                        maxCloseTime));
+    }
+
+    uint64_t const now = VirtualClock::to_time_t(systemNow);
+    uint64_t nextCloseTime = std::max(now, lastCloseTime + 1);
+
+    if (userSpecifiedCloseTime)
+    {
+        if (*userSpecifiedCloseTime > maxCloseTime)
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Specified close time past max ({})"),
+                            maxCloseTime));
+        }
+        if (*userSpecifiedCloseTime < nextCloseTime)
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Invalid manual close time {} (last "
+                                       "close time={}, now={})"),
+                            *userSpecifiedCloseTime, lastCloseTime, now));
+        }
+        nextCloseTime = *userSpecifiedCloseTime;
+    }
+    else
+    {
+        uint64_t const defaultManualCloseTimeInterval =
```

Comment:
just realized that this is a breaking change:
* existing `manualclose` users are going to see a clock from the 1970s instead of a current clock, this may break clients that are setting time bounds in their transactions
* a non breaking change would be to set `defaultNextCloseTime` to the current *real* clock's value (so calling `std::chrono::steady_clock::now()` directly)

Reply:
> calling `std::chrono::steady_clock::now()` directly

Why `std::chrono::steady_clock::now()` as opposed to `std::chrono::system_clock::now()`?  The latter is what `VirtualClock::system_now()` uses in `REAL_TIME`, and `VirtualClock::system_now()` is what `HerderImpl::triggerNextLedger()` uses.

Reply:
Done in 8f353399bbdac17b5645942260f5809e4cb7031f, though as I mentioned, I used `std::chrono::system_clock::now()`, not `std::chrono::steady_clock::now()`.  Please let me know if I'm misunderstanding something there.

Reply:
This is fine yes, I was worried that the clock would go backwards, but we're guarding against that now, so people would just have to wait a bit if this happens